### PR TITLE
Port EVM precompiles from Moonbeam

### PR DESCRIPTION
**Pull request by fgamundi**
Ports these EVM precompiled smart contracts from Moonbeam:

- pallet-balances
- batch
- call-permit
- xcm-utils

Also includes #10 

Depends on Frontier's https://github.com/paritytech/frontier/pull/1246

### CODE REVIEW

File: pallets/maintenance-mode/src/types.rs [hunk 1]
```diff
@@ -94,20 +94,11 @@ where
...
-		if Pallet::<T>::maintenance_mode() {
-			T::MaintenanceExecutiveHooks::pre_upgrade()
-		} else {
-			T::NormalExecutiveHooks::pre_upgrade()
-		}
-	}
-
-	#[cfg(feature = "try-runtime")]
-	fn post_upgrade(state: Vec<u8>) -> Result<(), sp_runtime::DispatchError> {
+	fn try_on_runtime_upgrade(checks: bool) -> Result<Weight, TryRuntimeError> {
```
Comment (girazoki) [inline pallets/maintenance-mode/src/types.rs:97]: Is this not part of master?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -242,6 +242,21 @@ jobs:
       CARGO_INCREMENTAL: "0"
       SCCACHE_CACHE_SIZE: "100GB"
     steps:
+      - name: Free Disk Space (Ubuntu)
```
Comment (girazoki) [inline .github/workflows/build.yml:245]: why this?

**Review by girazoki — APPROVED** [review]
Provided that frontier changes are approved, I think this PR looks good. @librelois what do you think?

File: pallets/maintenance-mode/src/types.rs (hunk 1 above)
Comment (fgamundi) [reply pallets/maintenance-mode/src/types.rs:97]: No, #10 was merged directly to the `moonbeam-polkadot-v1.1.0` branch. Maybe it's better to create a separate PR to cherrypick that into master and leave this PR only for the precompiles. What do you think?

File: .github/workflows/build.yml (hunk 2 above)
Comment (fgamundi) [reply .github/workflows/build.yml:245]: Not sure why it was needed, also part of #10. @rimbi can you give a bit of insight on this, please?

**Comment by fgamundi** [comment]
This became a bit stale. Closing in favor of #21

**Pull request closed without merging**